File: 2501~3000/2552. Count Increasing Quadruplets/main4.py
"""
Source: https://leetcode.com/problems/count-increasing-quadruplets/
Date: 2023/8/9
Skill:
Runtime: TLE
Memory Usage:
Time complexity:
Space complexity:
Constraints:
    4 <= nums.length <= 4000
"""

import math
from typing import List, Tuple, Optional
from collections import defaultdict, deque
from heapq import heapify, heappush, heappop, nsmallest, heappushpop
import bisect
import heapq
import functools


class Solution:
    def countQuadruplets(self, nums: List[int]) -> int:
        sz, res = len(nums), 0
        smaller_before, bigger_after = [[0 for _ in range(sz + 1)] for _ in range(sz + 1)], [[0 for _ in range(sz + 1)] for _ in range(sz + 1)]
        for i in range(sz):
            for j in range(1, sz + 1):
                if nums[i] < j:
                    smaller_before[i + 1][j] = smaller_before[i][j] + 1
                else:
                    smaller_before[i + 1][j] = smaller_before[i][j]
                if nums[sz - 1 - i] > j:
                    bigger_after[sz - 1 - i][j] = bigger_after[sz - i][j] + 1
                else:
                    bigger_after[sz - 1 - i][j] = bigger_after[sz - i][j]

        for i in range(1, sz - 2):
            for j in range(i + 1, sz - 1):
                if nums[i] > nums[j]:
                    res += smaller_before[i][nums[j]] * bigger_after[j][nums[i]]
        return res


if __name__ == "__main__":
    s = Solution()
    nums = [1, 3, 5, 2, 4]
    res = s.countQuadruplets(nums)
    print(res)

File: 2501~3000/2896. Apply Operations to Make Two Strings Equal/main4.cpp
/**
 * Source: tinyurl.com/ykgpb5um
 * Date: 2023/10/26
 * Skill:
 * Runtime: TLE
 * Memory Usage:
 * Time complexity: O(n)
 * Space complexity: O(n)
 * Constraints:
 *
 */

#include <vector>
#include <iostream>
#include <algorithm>
#include <map>
#include <set>
#include <unordered_set>
#include <queue>
#include <numeric>
#include <tuple>
#include <stack>
#include <bitset>
#include <functional>
#include <list>

using namespace std;


class Solution {
public:
    int minOperations(string s1, string s2, int x) {
        int cnt = 0, sz = s1.length();
        for (int i = 0; i < sz; i++) {
            if (s1[i] != s2[i]) cnt++;
        }
        if (cnt % 2 == 1) return -1;

        vector<vector<vector<int>>> dp(sz, vector<vector<int>>(sz, vector<int>(2, sz)));
        function<int(int, int, int)> dfs;
        dfs = [&](int idx, int free, int is_reversed) {
            if (idx == -1) {
                if (free > 0 || is_reversed > 0) return sz;
                return 0;
            }
            int &res = dp[idx][free][is_reversed];
            if (res < sz) return res;
            if ((s1[idx] == s2[idx]) == (is_reversed == 0)) {
                res = dfs(idx - 1, free, 0);
            } else {
                res = min(res, dfs(idx - 1, free, 1) + 1);
                res = min(res, dfs(idx - 1, free + 1, 0) + x);
                if (free > 0) res = min(res, dfs(idx - 1, free - 1, 0));
            }
            return res;
        };
        return dfs(sz - 1, 0, 0);
    }
};


int main() {
    Solution s;
}